Improve particle handler

gassmoeller opened the pull request:
This is work in progress and related to #1897. I moved most of the generic particle functionality into the particle handler class, but there is still a bug inside the serialization / transfer / deserialization of particles. I would already appreciate a review, since most stuff should be in its final state, I just need to find the problem.

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -128,6 +139,50 @@ namespace aspect
+         * This function allows to register three additional functions that are
+         * called every time a particle is transferred to another process
+         * (i.e. during sorting into cells, during ghost particle transfer, or
+         * during serialization of all particles.
```
missing `)`

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -128,6 +139,50 @@ namespace aspect
+         * right after its data block.
+         */
+        void
+        register_additional_store_load_functions(const std_cxx11::function<size_t ()> &size_callback,
```
std::size_t to be consistent with the other functions

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -128,6 +139,50 @@ namespace aspect
+         * the same for every particle, but it does not have to be the same for
+         * every serialization process (e.g. a serialization during particle
+         * movement might include temporary data, while a serialization after
+         * movement was finished does not need to transfer this data).
```
Out of curiosity, how is the callback supposed to know what is happening at the time it is called?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -128,6 +139,50 @@ namespace aspect
+         */
+        void
+        register_additional_store_load_functions(const std_cxx11::function<size_t ()> &size_callback,
+                                                 const std_cxx11::function<void *(const particle_iterator,
```
maybe make the iterator argument a reference? same for the third callback

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -128,6 +139,50 @@ namespace aspect
+                                                 const std_cxx11::function<void *(const particle_iterator,
+                                                                                  void *)> &store_callback,
+                                                 const std_cxx11::function<const void *(const particle_iterator,
+                                                                                        const void *)> &load_callback);
```
Here's some food for thought, not necessarily meant as a request to change things but for you to think about: In the `hp::DoFHandler` work, I've made these things more flexible by just letting the store functions return a `std::vector<char>` (could also have been a `std::string`) that the store function can size and fill as it pleases. I then let the place that calls all of these callbacks deal with the question of how to pack this data into buffers. That was easy in that case because it did not have to squeeze things into the p4est framework of attaching data to cells, but I found it kind of liberating to do it this way around because the store and load functions become so much simpler.  It required a little bit more logic in one central place, where I call these functions, but I thought that was worth it. Could something like this be done here as well?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -144,9 +199,60 @@ namespace aspect
+         * Get a map between subdomain id and the neighbor index. In other words
+         * the returned map answers the question: Given a subdomain id, which
+         * neighbor of the current processor's domain (in terms of a contiguous
+         * number from 0 to n_neighbors) owns this subdomain?
```
so is the value of the map then a number between zero and `n_mpi_processes(...)`? (it's not an MPI rank, I think.) maybe say this -- I had to read this sentence a few times before I understood what the function does.

By the way do you need this function to be public?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -144,9 +199,60 @@ namespace aspect
+         * 'particles' member variable in this function.
+         */
+        void
+        sort_particles_in_subdomains_and_cells();
```
the documentation here does not match the signature any more, as it talks about `particles_to_sort`.

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -144,9 +199,60 @@ namespace aspect
+
+
+        /**
+         * Exchanges all particles that live in cells adjacent to ghost cells
```
that live in cells that are ghost cells on other processors?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -171,13 +277,33 @@ namespace aspect
+        /**
+         * Const access to particles in this world.
+         * TODO: This function needs to go to not expose internal structure.
+         * This can only be done once World no longer uses this function.
```
Are you worried about exposing this function because it shows the underlying data structure when you carefully hide the data structure for the locally owned particles?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -171,13 +277,33 @@ namespace aspect
         SmartPointer<const parallel::distributed::Triangulation<dim,spacedim>,ParticleHandler<dim,spacedim> > triangulation;
 
         /**
+         * Address of the triangulation to work on.
```
-> mapping

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -211,6 +344,58 @@ namespace aspect
+                                   void *)> store_callback;
+
+        /**
+         * @param size_callback A function that is called when serializing
```
remove the `@param`

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -237,6 +422,50 @@ namespace aspect
+        void
+        send_recv_particles(const std::vector<std::vector<particle_iterator> > &sent_particles,
+                            std::multimap<types::LevelInd,Particle <dim> >     &received_particles,
+                            const std::vector<std::vector<active_cell_it> > &sent_cells = std::vector<std::vector<active_cell_it> > ());
```
can you also document the last argument?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -237,6 +422,50 @@ namespace aspect
 
         /**
+         * Transfer particles that have crossed subdomain boundaries to other
+         * processors. The transfer occurs in two steps. As a first step all
```
I'd drop the explanation of *how* the function works from the documentation. What's relevant here is *what* the function does. I say that because I think I'll have comments on how this is implemented once I get to this part of the patch :-)

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -237,6 +422,50 @@ namespace aspect
+         * @p received_particles vector.
+         *
+         * @param [in] sent_particles All particles that should be sent and
+         * their new subdomain_ids are in this map.
```
maybe call it `particles_to_send`? how are the nested vectors organized, i.e., what do the indices mean?

bangerth commented on include/aspect/particle/particle_handler.h:
```diff
@@ -237,6 +422,50 @@ namespace aspect
+        /**
+         * Called by listener functions from Triangulation for every cell
+         * before a refinement step. All particles have to be attached to their
+         * element to be sent around to the new cell/processes.
```
element -> cell

bangerth commented on include/aspect/simulator.h:
```diff
@@ -1391,6 +1390,8 @@ namespace aspect
       DoFHandler<dim>                                           dof_handler;
 
+      Postprocess::Manager<dim>                                 postprocess_manager;
+
```
any particular reason to move the place where we declare this manager (rather than sticking with all of the other managers)? is it because you need to have access to the `dof_handler` when you initialize it?

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+
+    template <int dim, int spacedim>
+    void
+    ParticleHandler<dim,spacedim>::exchange_ghost_particles()
```
instead of reading through this function -- could you instead use the new `parallel::GridTools::exchange_...` function Timo wrote at the hackathon for what you're doing here? that would seem simpler. Since you're not yet working with deal.II dev here, you may have to copy that function from deal.II but that shouldn't be terribly difficult.

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+    {
+      // Determine the communication pattern
+      const std::vector<types::subdomain_id> neighbors (triangulation->ghost_owners().begin(),
+                                                        triangulation->ghost_owners().end());
```
that's dangerous -- it only works if `ghost_owners()` returns a reference to whatever it is storing internally. If it is returning a *copy* of the object, then taking `begin` and `end` will yield pointers that point to separate objects.

It may be safer to copy what that function returns.

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        MPI_Irecv(&(n_recv_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i]));
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        MPI_Isend(&(n_send_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i+1]));
```
do the sends first then the receives

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+        MPI_Irecv(&(n_recv_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i]));
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        MPI_Isend(&(n_send_data[i]), 1, MPI_INT, neighbors[i], 0, mpi_communicator, &(n_requests[2*i+1]));
+      MPI_Waitall(2*n_neighbors,&n_requests[0],MPI_STATUSES_IGNORE);
```
if you enclose lines 689-695 into braces, you can let the `n_requests` variable die right here where it is used last

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+      for (unsigned int i=0; i<n_neighbors; ++i)
+        if (n_send_data[i] > 0)
+          {
+            MPI_Isend(&(send_data[send_offsets[i]]), n_send_data[i], MPI_CHAR, neighbors[i], 1, mpi_communicator,&(requests[send_ops+recv_ops]));
```
same here -- do the sends before the receives

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+      // Put the received particles into the domain if they are in the triangulation
+      const void *recv_data_it = static_cast<const void *> (&recv_data.front());
+
+      while (reinterpret_cast<std::size_t> (recv_data_it) - reinterpret_cast<std::size_t> (&recv_data.front()) < total_recv_data)
```
could you avoid the `reinterpret_cast` by instead doing something like
```
  while (rec_data_it < static_cast<void*>(&recv_data.back()))
```
or similar?

bangerth commented on source/particle/particle_handler.cc:
```diff
@@ -222,6 +261,661 @@ namespace aspect
+
+              const CellId::binary_type cellid = cell->id().template to_binary<dim>();
+              memcpy(data, &cellid, cellid_size);
+              data = static_cast<char *>(data) + cellid_size;
```
it would be nicer to use the boost serialization facilities here, but that can wait for another day.

bangerth commented on source/particle/world.cc:
```diff
@@ -165,7 +112,7 @@ namespace aspect
                              "particles, but the parameter 'Update ghost particles' has not "
                              "been set, therefore no ghost particles are available."));
-      return ghost_particles;
+      return particle_handler->get_ghost_particles();
```
we should eventually convert the places that use these ghost particles to call something like `particle_handler.begin_ghost()` and `particle_handler.end_ghost()`, and then use iterators as well. This would allow you to hide the underlying data structure the same way you do for the locally owned particles. 

Not sure whether that's easy to do at the moment, though.

gassmoeller commented:
I think I addressed most comments, lets see what the tester says. If the tests pass I will rebase and squash. To your questions:

> Out of curiosity, how is the callback supposed to know what is happening at the time it is called?

I removed the confusing comment for now. At the moment the function is only called in one place anyway, and we can think about a better solution when it is used in more places.

> Here's some food for thought, not necessarily meant as a request to change things but for you to think about: In the hp::DoFHandler work, I've made these things more flexible by just letting the store functions return a std::vector<char> (could also have been a std::string) that the store function can size and fill as it pleases. I then let the place that calls all of these callbacks deal with the question of how to pack this data into buffers. That was easy in that case because it did not have to squeeze things into the p4est framework of attaching data to cells, but I found it kind of liberating to do it this way around because the store and load functions become so much simpler. It required a little bit more logic in one central place, where I call these functions, but I thought that was worth it. Could something like this be done here as well?

I am a bit worried about the speed of your suggestion. Usually this function is called once per transferred particle (= for every particle during mesh refinement), and every call to this function serializes a very small amount of data (often <= 50 bytes per particle). Copying 50 bytes is likely much faster than allocating an std::vector and copying the data. Of course one could try if it matter at all compared to the rest of the algorithm, but for now I would like to keep the interface constant.

> Are you worried about exposing this function because it shows the underlying data structure when you carefully hide the data structure for the locally owned particles?

Exactly, that is what the TODO was supposed to say. I will remove all usage of this function in follow-up PRs, and then it can be removed.

> any particular reason to move the place where we declare this manager (rather than sticking with all of the other managers)? is it because you need to have access to the dof_handler when you initialize it?

I need access to the mapping (or precisely: I need the mapping to be constructed first and destructed last so that the subscriptor pointer is correctly constructed and destructed). Alternatively I could leave the manager in its old position and move the mapping further up, would you prefer that?

> I don't recall, but do you say in the documentation that you are also initializing the particle in this function? Out of curiosity, is there a function that inserts a whole bunch of particles at the same time? Wasn't that one of the things that's important to do in bulk?

I now copy the particle properties, which is what a user would expect, but I also mention it in the documentation. 

Currently there is no function that enters many particles at once, but I do not use this function yet anyway (the particle generators still access the data structure directly). I will fix that in a follow-up PR that deals with all of our interface changes at once.

> instead of reading through this function -- could you instead use the new parallel::GridTools::exchange_... function Timo wrote at the hackathon for what you're doing here? that would seem simpler. Since you're not yet working with deal.II dev here, you may have to copy that function from deal.II but that shouldn't be terribly difficult.

I would prefer to keep that for a follow-up PR. There are several functions that can be replaced by deal.II functionality, but if I start mixing up moving functionality and replacing functionality that is simply a recipe for complete chaos.

> do the sends first then the receives

Do you know this is better, or is this just your gut feeling? All examples of Isend and Irecv I found are receives first then sends. And http://www.cas.mcmaster.ca/~nedialk/COURSES/mpi/Lectures/lec2_1.pdf claims that `A communication will generally have lower overhead if a receive buffer is already posted when a sender initiates communication`. I have not measured that though.

> could you avoid the reinterpret_cast by instead doing something like `while (rec_data_it < static_cast<void*>(&recv_data.back()))` or similar?

Maybe, but I remember that I played around with this casts already quite a bit (they changed several times over the iterations of the original merge), and I do not think this PR is the right time to try this.

> we should eventually convert the places that use these ghost particles to call something like particle_handler.begin_ghost() and particle_handler.end_ghost(), and then use iterators as well. This would allow you to hide the underlying data structure the same way you do for the locally owned particles.

I completely agree. That is something I will address in a follow-up PR.

bangerth commented:
>     Here's some food for thought, not necessarily meant as a request to
>     change things but for you to think about: In the hp::DoFHandler
>     work, I've made these things more flexible by just letting the store
>     functions return a std::vector (could also have been a std::string)
>     that the store function can size and fill as it pleases. I then let
>     the place that calls all of these callbacks deal with the question
>     of how to pack this data into buffers. That was easy in that case
>     because it did not have to squeeze things into the p4est framework
>     of attaching data to cells, but I found it kind of liberating to do
>     it this way around because the store and load functions become so
>     much simpler. It required a little bit more logic in one central
>     place, where I call these functions, but I thought that was worth
>     it. Could something like this be done here as well?
> 
> I am a bit worried about the speed of your suggestion. Usually this 
> function is called once per transferred particle (= for every particle 
> during mesh refinement), and every call to this function serializes a 
> very small amount of data (often <= 50 bytes per particle). Copying 50 
> bytes is likely much faster than allocating an std::vector and copying 
> the data. Of course one could try if it matter at all compared to the 
> rest of the algorithm, but for now I would like to keep the interface 
> constant.

Yes, that makes sense.


>     any particular reason to move the place where we declare this
>     manager (rather than sticking with all of the other managers)? is it
>     because you need to have access to the dof_handler when you
>     initialize it?
> 
> I need access to the mapping (or precisely: I need the mapping to be 
> constructed first and destructed last so that the subscriptor pointer is 
> correctly constructed and destructed). Alternatively I could leave the 
> manager in its old position and move the mapping further up, would you 
> prefer that?
> 
>     I don't recall, but do you say in the documentation that you are
>     also initializing the particle in this function? Out of curiosity,
>     is there a function that inserts a whole bunch of particles at the
>     same time? Wasn't that one of the things that's important to do in bulk?
> 
> I now copy the particle properties, which is what a user would expect, 
> but I also mention it in the documentation.
> 
> Currently there is no function that enters many particles at once, but I 
> do not use this function yet anyway (the particle generators still 
> access the data structure directly). I will fix that in a follow-up PR 
> that deals with all of our interface changes at once.
> 
>     instead of reading through this function -- could you instead use
>     the new parallel::GridTools::exchange_... function Timo wrote at the
>     hackathon for what you're doing here? that would seem simpler. Since
>     you're not yet working with deal.II dev here, you may have to copy
>     that function from deal.II but that shouldn't be terribly difficult.
> 
> I would prefer to keep that for a follow-up PR. There are several 
> functions that can be replaced by deal.II functionality, but if I start 
> mixing up moving functionality and replacing functionality that is 
> simply a recipe for complete chaos.

Yes, indeed!


>     do the sends first then the receives
> 
> Do you know this is better, or is this just your gut feeling? All 
> examples of Isend and Irecv I found are receives first then sends. And 
> http://www.cas.mcmaster.ca/~nedialk/COURSES/mpi/Lectures/lec2_1.pdf 
> <http://www.cas.mcmaster.ca/%7Enedialk/COURSES/mpi/Lectures/lec2_1.pdf> 
> claims that |A communication will generally have lower overhead if a 
> receive buffer is already posted when a sender initiates communication|. 
> I have not measured that though.

You're completely correct. I don't recall the context of my request, but 
your suggestion is entirely reasonable.

bangerth commented:
Thanks for working through this! I look forward to the next steps!
